File: app/src/main/java/org/miaowo/miaowo/transition/MoveTransition.java
package org.miaowo.miaowo.transition;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.transition.Transition;
import android.support.transition.TransitionValues;
import android.view.ViewGroup;

public class MoveTransition extends Transition {
    @Override
    public void captureEndValues(@NonNull TransitionValues transitionValues) {
        transitionValues.values.put("x", transitionValues.view.getTranslationX());
        transitionValues.values.put("y", transitionValues.view.getTranslationY());
    }

    @Override
    public void captureStartValues(@NonNull TransitionValues transitionValues) {
        transitionValues.values.put("x", transitionValues.view.getTranslationX());
        transitionValues.values.put("y", transitionValues.view.getTranslationY());
    }

    @Nullable
    @Override
    public Animator createAnimator(@NonNull ViewGroup sceneRoot, @Nullable TransitionValues startValues, @Nullable TransitionValues endValues) {
        if (startValues == null || endValues == null) return null;
        float startX, startY, endX, endY;
        startX = startValues.values.get("x") == null ? 0 : (float) startValues.values.get("x");
        endX = endValues.values.get("x") == null ? 0 : (float) endValues.values.get("x");
        startY = startValues.values.get("y") == null ? 0 : (float) startValues.values.get("y");
        endY = endValues.values.get("y") == null ? 0 : (float) endValues.values.get("y");
        return ObjectAnimator.ofPropertyValuesHolder(startValues.view,
                PropertyValuesHolder.ofFloat("translationX", startX, endX),
                PropertyValuesHolder.ofFloat("translationY", startY, endY));
    }
}

File: app/src/main/java/org/miaowo/miaowo/other/MoveTransition.kt
package org.miaowo.miaowo.other

import android.animation.Animator
import android.animation.ObjectAnimator
import android.animation.PropertyValuesHolder
import android.support.transition.Transition
import android.support.transition.TransitionValues
import android.view.ViewGroup

/**
 * MiaoFragment 界面动画集合中的移动动画
 * 为了兼容老版本
 */
class MoveTransition : Transition() {
    override fun captureEndValues(transitionValues: TransitionValues) {
        transitionValues.values.put("x", transitionValues.view.translationX)
        transitionValues.values.put("y", transitionValues.view.translationY)
    }

    override fun captureStartValues(transitionValues: TransitionValues) {
        transitionValues.values.put("x", transitionValues.view.translationX)
        transitionValues.values.put("y", transitionValues.view.translationY)
    }

    override fun createAnimator(sceneRoot: ViewGroup, startValues: TransitionValues?, endValues: TransitionValues?): Animator? {
        if (startValues == null || endValues == null) return null
        val startX = startValues.values["x"] as Float
        val startY = startValues.values["y"] as Float
        val endX = endValues.values["x"] as Float
        val endY = endValues.values["y"] as Float
        return ObjectAnimator.ofPropertyValuesHolder(startValues.view,
                PropertyValuesHolder.ofFloat("translationX", startX, endX),
                PropertyValuesHolder.ofFloat("translationY", startY, endY))
    }
}
